blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,562 @@
+
+// Configure the OTLP exporter to send data to your OpenTelemetry backend
+const config = { url: 'https://otel.highlight.io:4317' }
+const sdk = new NodeSDK({
+  autoDetectResources: true,
+  resourceDetectors: [processDetectorSync],
+  resource: new Resource({
+    [SEMRESOURCENAME.SERVICE_NAME]: 'nextjs-app',
+    'highlight.project_id': '<YOUR_PROJECT_ID>',
+  }),
+  spanProcessors: [spanProcessor],
+  logRecordProcessors: [logProcessor],
+  metricReader: metricReader,
+  traceExporter: exporter,
+  contextManager: new AsyncLocalStorageContextManager(),
+  sampler: new AlwaysOnSampler(),
+  instrumentations: [new HttpInstrumentation(), new FetchInstrumentation()],
+})
+
+sdk.start();
+console.log('OpenTelemetry initialized');
+```
+
+To trigger this file to run when the app starts, you can invoke it from the Next.js magic `instrumentation.ts` file.
+
+```typescript
+export function register() {
+  await import('./otel');
+}
+```
+
+The [instrumentation.ts file](https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation) is automatically detected by Next.js and will run when the app starts. Before Next.js 15, the instrumentation is experimental, so you will have to enable it explicitly:
+
+```typescript
+module.exports = {
+  experimental: {
+    instrumentationHook: true,
+  },
+};
+```
+
+
+### Configuring Tracing
+
+With the SDK configured, your application will start to export the telemetry data using the exporters defined.
+However, you may wonder what data is being captured without any explicit code added.
+
+[Next.js has built-in OpenTelemetry spans](https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry#default-spans-in-nextjs)
+for various parts of the application, including:
+- API routes (pages/api or app/api)
+- Page router (Pages Directory)
+- App router (App Directory)
+
+Some top-level spans are emitted out-of-the-box, while others can be turned on by turning on verbose logging:
+
+```bash
+NEXT_OTEL_VERBOSE=1
+```
+
+Setting the `NEXT_OTEL_VERBOSE` environment variable will emit additional traces that give you more granularity of the code execution.
+
+For example, here's a flame graph visualization of a trace without verbose tracing, `NEXT_OTEL_VERBOSE=0`:
+
+![](/images/blog/nextjs-otel/not-verbose.png)
+
+And here's the same trace with verbose tracing enabled, `NEXT_OTEL_VERBOSE=1`:
+
+![](/images/blog/nextjs-otel/verbose.png)
+
+Let's go through some examples of the data that can be captured.
+
+![](/images/blog/nextjs-otel/trace.png)
+
+In the image above, you can see the trace start with an api route request that is piped through Next.js to the API handler.
+We also see a custom span that wraps an ourgoing API request to another service. Because we set up auto-instrumentation, 
+we capture the `fetch` call automatically, and can even propagate the trace context to the backend service.
+
+Here's a list of the top-level spans that are captured automatically by Next.js:
```

Comment:
Nit, but the formatting is a bit wonky in this section. WDYT about making the span name bold and then putting the description on the same line? e.g.

* **BaseServer.handleRequest** - Named `[http.method] [next.route]`, [the root span for each incoming request to your Next.js application](https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry#httpmethod-nextroute).